### ISSUE TITLE
Adding some way to call intero externally

### ISSUE THREAD

**Issue by lwm** [commercialhaskell/intero#390]
Follows the (short!) discussion over at https://github.com/parsonsmatt/intero-neovim/issues/18.

Willing to do all sorts of strange things to make this possible. Just point in the direction.

**Comment by rlupton20** [comment]
Looks like what you want is quite generic. `emacs` uses, for instance, comint mode: https://www.emacswiki.org/emacs/ComintMode

Surely `neovim` has something for this?

**Comment by lwm** [comment]
@rlupton20, could be wrong but I think comint mode is missing in vim!

**Comment by rlupton20** [comment]
@lwm sorry, I meant does neovim ( / vim) have an equivalent of comint mode? At the very least something which let's you dispatch commands to a background process and get back the output.

Sorry if I'm being unhelpful, I had a skim of the issue referenced but know very little of that ecosystem.

**Comment by rdnetto** [comment]
To add some context:
* Neovim definitely has the ability to dispatch commands in the background and get the output from them
* intero-neovim keeps the intero process open in a buffer, which it can read to / write from independently of whether the user has that buffer open/visible
* this means that the only way to interact with that intero process is through Neovim functions
* Neomake, the standard way of implementing flychecking in Neovim, is designed to execute processes/shell scripts rather than Neovim functions
* we definitely want to call out to an existing Intero process rather than a new one, as otherwise we lose incremental recompilation

So adding the ability to call out to an existing Intero process from the command line is seen as one path to integrating it with Neomake. (The other option is to change Neomake to support using an existing process instead.)

There's a related issue, which is that currently to implement any command intero-neovim must execute it in the buffer, then parse the output. This exposes the implementation details to the user and is quite fragile, as any unexpected output can result in cryptic error messages. It also pollutes the user's history in Intero. If we had a RPC mechanism to use instead of this, then we could have multiple clients (a visible one for the user and a hidden one for commands / flychecking) connected to the same Intero process.

[Someone once forked Intero](https://github.com/myfreeweb/intero.nvim) to add a RPC mechanism like this, but the changes were never upstreamed (I don't know why) and the repo has not been maintained in a year.

So tldr;
* adding an RPC mechanism to Intero would solve several problems, including flychecking
* the motivating problem (flychecking) has an alternative solution (modifying Neomake) of unknown difficulty
* this is potentially useful to other consumers of Intero as well

**Comment by lwm** [comment]
> Sorry if I'm being unhelpful

No, all help is much appreciated!

Thank you very much @rdnetto! Great summary.

> So adding the ability to call out to an existing Intero process from the command line is seen as one path to integrating it with Neomake. (The other option is to change Neomake to support using an existing process instead.)
> ...
> * the motivating problem (flychecking) has an alternative solution (modifying Neomake) of unknown difficulty

/cc @blueeyed, (hi from the internet!), a question about Neomake! You can see from the thread but I wanted to integrate a linter to Neomake and found that it has no command line interface, only a running process. Is there support in Neomake (or planned) to use processes and not a command line interface for calling linters? Thanks! (Hope that makes sense, I never really knew what I was talking about when it came to Vim internals)

**Comment by blueyed** [comment]
@lwm 
What is "a running process" in this case?  A daemon?
Can you interact with it using `curl` / `JSON`?  (that would be possible with Neomake)
You can also have a maker with a `get_list_entries` method that can then just do anything (without using Neomake's process job handling).

**Comment by lwm** [comment]
Oh, forgot to reply here, sorry! It's been done in https://github.com/parsonsmatt/intero-neovim/issues/18#issuecomment-310234636. Thanks @rdnetto!
